### ISSUE TITLE
CheckSite Error: The 'Domain'='localhost:62017' part of the cookie is invalid.

### ISSUE THREAD

**Issue by medmondson** [RickStrahl/WestWindWebSurge#44]
I've recorded a request in Fiddler which returns the correct response whenever I reissue it.  I've then copied the exact same request into WebSurge through the following method:

- Taking it from Inspectors - Raw in Fiddler
- Copy that into a new request and using the edit button pasting the above request into notepad that appears, then save.

When I test this in WebSurge I get the following error:

![websurgeerror](https://user-images.githubusercontent.com/1955888/29877795-7f1a9d00-8d98-11e7-8f8a-d60c477a9921.png)

The raw request in WebSurge (identical to Fiddler)

```
GET http://localhost:62017/MeterRead HTTP/1.1
Host: localhost:62017
Connection: keep-alive
Cache-Control: max-age=0
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/60.0.3112.113 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Referer: http://localhost:62017/
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.8
Cookie: cookieconsent_status=dismiss; __RequestVerificationToken=yHS2-1NZXT-LDUwSlg2RfFGHOKOVO6Mmgg7pxxkiUoDjMxQeKuqz4Ui3U0G2JEnwxXMNXsuMO_ihNV_biLyLkpFuW181; ASP.NET_SessionId=omvkkxc0ahpasdko3sgieafd; .AspNet.ApplicationCookie=Y1PPUa7MlloTd4ZzmlFGX94f2eJm_T2z9FKKEXOzABK5eLyV5MH3eI2rncE5hsG3shUPmg3eUCbwh3ESWsRIMA5M6VZizKT2t0XhK8C1pjnSnth5F81qtIkIpOfe8Vgx0pXCHaTqpfvlQrEWIG7CfZ0mZVO-53nfZc5uKCG-DuyZKkLsTgqCngBmowqExE5PIMzuEtewYVe4iawmLuH3H0ZHwojeRTPGrSprEmnrucGm666BJM3XntreFI-1xrsHFfuopz1rC9AINwrl1buxtURhNTCNlemzrMLaHOlRggowFC_93mFEjL26G6O3wcFj8wIvxfs1bz-J9_w39FFlEFbdYp7uHataeEFfAII8rtbZ8UM6tSvlXBi5erE8OoizqDWnazlSHsSi8J207oDyl4ok0vjGKcvPGKvcuvSxw2MYazQLxC8hkwMrj52-V-jWcbgyhbHPGJiAQ1ucUVt3cYYP3qI2s4El7LO_3vqhWr5ULCJzo0FpVOGl4IzODU45ZpFz4el0M59FF2CuiBsOG4ppNO-JuWL9vrmuVERIqqrV1J5e8hPfeNIbUKNLtkfk4gmE_Z962Mw_W1Q5nMO5xn6K6b4; _gat=1; _ga=GA1.1.1500294237.1499094041; _gid=GA1.1.259567524.1504004538
```

Fiddler returns HTTP 200 each time with the correct content, therefore I suspect we have an issue,

**Comment by RickStrahl** [comment]
Most likely a cookie issue. Check the `TrackPerSessionCookie` setting. If on - it'll use a previous captured cookie for this domain when you reissue you the request. If you want to use the cookie in the request rather than the captured cookie turn the `TrackPerSessionCookie` option `false`.

FWIW, the default of this setting will change back to `false` in future updates.

**Comment by medmondson** [comment]
For some reason this has started working - although I found I had to log-in via my browser and capture the cookie that way then stick it into 'ReplaceCookieValue'.

As an aside I found the error confusing, it's not clear if the error is from the server (and the response has been wrapped into this message) or from WebSurge itself.  If from WebSurge I'd be questioning why it needs to intervene in the communication.

Perhaps I'm missing something but I think there's value in working on the clarity here - something I'd be happy to help with?

**Comment by medmondson** [comment]
I've managed to replicate it:

- Capture some traffic (a GET with a multiple cookies)
- Test this request - I consistently get the above error.

To resolve I have to
- Set 'TrackPerSessionCookies' to false
- Copy the cookies from the request and put into 'ReplaceCookieValue'

**Comment by RickStrahl** [comment]
I think the error comes from the HTTP stack. When `SetTrackPerSessionCookies=true` I use a CookieContainer and the CookieContainer validates cookies. So if there's an invalid cookie I believe that the container complains.

Ultimately if you need a specific cookie to be used you need to turn off `TrackPerSessionCookies` in which case cookies are just a raw HTTP header and go straight through.

Is this confusing? Yeah probably - but frankly I don't know how to make this better. Cookie handling in a disconnected environment where you run one way interactively and another when running a session is difficult to consolidate. If you have any ideas, I'd love to hear it and we can see about adding/modifying.

A related topic is managing multiple users and sessions which is still being fleshed out. From the docs:  

* [Setting up User Sessions](https://websurge.west-wind.com/docs/_4ym15ftd2.htm)

**Comment by RickStrahl** [comment]
This has been fixed for v1.9 and later. There were actually a number of errors in the Cookie parsing, resulting in non-trimmed cookie names and duplicated cookies being added to the cookie container. 

Fixed. 528dde0ada5a0f2feaa047cd82bf5b2072e257a2
